daemon/graphdriver/copy/copy.go

```diff
@@ -9,32 +9,94 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/docker/docker/pkg/parsers/kernel"
+
+	"sync/atomic"
+
 	"github.com/docker/docker/pkg/pools"
 	"github.com/docker/docker/pkg/system"
 	rsystem "github.com/opencontainers/runc/libcontainer/system"
 	"golang.org/x/sys/unix"
 )
 
-type copyFlags int
+var copyFileRangeDisabled int32
+
+// The maximum amount of bytes Linux will splice
+const maxRwCount = 0x7ffff000
+
+// Mode indicates whether to use hardlink or copy content
+type Mode int
 
 const (
-	copyHardlink copyFlags = 1 << iota
+	// Content creates a new file, and copies the content of the file
+	Content Mode = iota
+	// Hardlink creates a new hardlink to the existing file
+	Hardlink
 )
 
-func copyRegular(srcPath, dstPath string, mode os.FileMode) error {
+func isCopyFileRangePosssible(srcDir, dstDir string) bool {
+	if copyFileRangeDisabled != 0 {
+		return false
+	}
+	// The copy_file_range syscall was introduced in kernel 4.5
+	v, err := kernel.GetKernelVersion()
+	if err != nil {
+		return false
+	}
+	if kernel.CompareKernelVersion(*v, kernel.VersionInfo{Kernel: 4, Major: 5, Minor: 0}) < 0 {
+		atomic.StoreInt32(&copyFileRangeDisabled, 1)
```

Comment:
Atomic ops can be fairly expensive on many platforms (cross cpu sync/locks), is full atomicity strictly required here? It doesn't seem like the end of the world if a few parallel copy attempts try the fast path and hit the fallback and make multiple assignments to a regular `bool` before it becomes globally visible (unless the go memory model is far more relaxed than I expect).

Also I think you would have to use `atomic.LoadInt32` in the check in order to get the benefits of using this atomic store, otherwise the read is not ordered wrt the store. Witht that then you would be doing an atomic read on every copy which adds to the overhead even more.

Reply:
I'm fairly sure there's no guarantees about single (simultaneous) goroutine guarantees about access to the graphdriver. Fairly sure that not using atomic here will result in violating the Go memory model. Compared to all the syscalls / IOPs here, I think the atomic ops are cheap here.

Reply:
If this is set 1 time during startup, there is no need for atomic.

Reply:
@cpuguy83 Even better, just do it in `init()` (see https://github.com/moby/moby/pull/34670#discussion_r136038882). Also, I'm a bit confused why this has been moved to a separate package -- does any other graphdriver need this?

Reply:
I don't think so. `vfs` could use it, but it currently uses `archive.CopyWithTar`

Reply:
1) I'm intending on using this for VFS eventually, so I split it out.
2) Testing it when it's an independent package is easier.